server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.OpenSearchException;
+import org.opensearch.action.search.SearchShardIterator;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.metadata.WeightedRoutingMetadata;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.index.shard.ShardId;
+import org.opensearch.search.SearchShardTarget;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+/**
+ * This class contains logic to find next shard to retry search request in case of failure from other shard copy.
+ * This decides if retryable shard search requests can be tried on shard copies present in data
+ * nodes whose attribute value weight for weighted shard routing is set to zero.
+ */
+
+public class FailOpenRouting {
+
+    private static final Logger logger = LogManager.getLogger(FailOpenRouting.class);
+
+    /* exception causing failure for previous shard copy */
+    private Exception exception;
+
+    private ClusterState clusterState;
+
+    public FailOpenRouting(Exception e, ClusterState clusterState) {
+        this.exception = e;
+        this.clusterState = clusterState;
+    }
+
+    /**
+     * *
+     * @return true if exception is due to cluster availability issues
+     */
+    private boolean isInternalFailure() {
+        if (exception instanceof OpenSearchException) {
+            return ((OpenSearchException) exception).status().getStatus() / 100 == 5;
+        }
+        return false;
+    }
```

Comment:
Please lets be more specific here, prefer using something like RestStatus#INTERNAL_SERVER_ERROR